src/Pos/Delegation/Logic.hs

```diff
@@ -170,40 +134,305 @@ initDelegation = do
+    => DelegationStateAction m a -> m a
+runDelegationStateAction action = do
+    var <- askDelegationState
+    RWV.modify var $ \startState -> swap <$> runStateT action startState
+
+-- | Invalidates proxy caches using built-in constants.
+invalidateProxyCaches :: (Monad m) => UTCTime -> DelegationStateAction m ()
+invalidateProxyCaches curTime = do
+    dwMessageCache %=
+        filterLRU (\t -> addUTCTime (toDiffTime messageCacheTimeout) t > curTime)
+    dwConfirmationCache %=
+        filterLRU (\t -> addUTCTime (toDiffTime lightDlgConfirmationTimeout) t > curTime)
+  where
+    toDiffTime (t :: Integer) = fromIntegral t
+
+-- Retrieves psk certificated that have been accumulated before given
+-- block. The block itself should be in DB.
+getPSKsFromThisEpoch
+    :: forall ssc m.
+       DB.MonadBlockDB ssc m
+    => HeaderHash -> m [ProxySKHeavy]
+getPSKsFromThisEpoch tip =
+    concatMap (either (const []) (getDlgPayload . view mainBlockDlgPayload)) <$>
+        (DB.loadBlocksWhile @ssc) isRight tip
+
+-- Copied from 'these' library.
+data These a b = This a | That b | These a b
+    deriving (Eq, Show, Generic)
+
+instance (B.Buildable a, B.Buildable b) => B.Buildable (These a b) where
+    build (This a)    = bprint ("This {"%build%"}") a
+    build (That a)    = bprint ("That {"%build%"}") a
+    build (These a b) = bprint ("These {"%build%", "%build%"}") a b
+
+-- u → (Maybe d₁, Maybe d₂): u changed delegate from d₁ (or didn't
+-- have one) to d₂ (or revoked delegation). These a b ≃ (Maybe a,
+-- Maybe b) w/o (Nothing,Nothing).
+type TransChangeset = HashMap PublicKey (These PublicKey PublicKey)
+type ReverseTrans = HashMap PublicKey (HashSet PublicKey, HashSet PublicKey)
+
+-- WHENEVER YOU CHANGE THE FUNCTION, CHECK DOCUMENTATION CONSISTENCY! THANK YOU!
+--
+-- Takes a set of dlg edge actions to apply and returns compensations
+-- to dlgTrans and dlgTransRev parts of delegation db. Should be
+-- executed under shared Gstate DB lock.
+calculateTransCorrections
+    :: forall m.
+       (MonadDBPure m, WithLogger m)
+    => HashSet GS.DlgEdgeAction -> m SomeBatchOp
+calculateTransCorrections eActions = do
+    -- Get the changeset and convert it to transitive ops.
+    changeset <- transChangeset
+    let toTransOp iPk (This _)      = GS.DelTransitiveDlg iPk
+        toTransOp iPk (That dPk)    = GS.AddTransitiveDlg iPk dPk
+        toTransOp iPk (These _ dPk) = GS.AddTransitiveDlg iPk dPk
+    let transOps = map (uncurry toTransOp) (HM.toList changeset)
+
+    -- Once we figure out this piece of code works like a charm we
+    -- can delete this logging.
+    logDebug $ sformat ("Trans changeset: "%mapJson) $ HM.toList changeset
+
+    -- Bulid reverse transitive set and convert it to reverseTransOps
+    let reverseTrans = buildReverseTrans changeset
+
+    reverseOps <- forM (HM.toList reverseTrans) $ \(k, (ad0, dl0)) -> do
+        let ad = HS.toList ad0 -- view patterns are for the weak ones
```

Comment:
Why are you converting everything to list, then do operations like `intersect` and `\\` (which should be cheaper for set) and then convert back to set?